Automattic/woocommerce-subscriptions-core
# Added notification emails and basic scheduling logic

First POC that contributes towards https://github.com/Automattic/woocommerce-subscriptions-core/issues/645

## Description

This is a WIP to add notifications to Subscriptions Core. 

Currently, it 
- adds simple email templates (placeholders until we [figure out the specifics](https://woomarketingstudio.wordpress.com/2024/07/10/studio-request-notify-users-about-new-minimum-required-php-woocommerce-and-wordpress-versions-for-all-somewherewarm-extensions/)), 
- adds email-related classes, 
- registers these classes, 
- adds simple scheduling logic for actions that should trigger the email with an offset (3 days)

Mostly, it's inspired by code that's already in Subscriptions Core. 

### Notification Types Added

- Auto Renewal
  - scheduled {{time_offset}} ahead of `$subscription->get_date( 'next_payment' )` for automatically renewing subscriptions (`$subscription->is_manual()` is `false`)
- Manual Renewal
  - scheduled {{time_offset}} ahead of `$subscription->get_date( 'next_payment' )` manual subscriptions (`$subscription->is_manual()` is `true`)
- Trial Expiry
  - scheduled {{time_offset}} ahead of `$subscription->get_date( 'trial_end' )` manual  subscriptions
- Subscription End
  - scheduled {{time_offset}} ahead of `$subscription->get_date( 'end' )` manual  subscriptions



### Open Questions

Some questions that crossed my mind while implementing this:
- We need to decide where to put filters/actions to allow extensibility
- I could have added the schedules to `WCS_Action_Scheduler`, but it seemed more appropriate to create a new class for notifications (`WCS_Action_Scheduler_Customer_Notifications `), otherwise the scheduling logic might be a bit complex. It should be quite easy to change this, though.
- The scheduling logic currently uses the state of the subscription object instead of deciding depending on which date/status is being updated. It was easier to grasp for me, given my inexperience with Subscriptions, but again, it should be possible to convert this to an approach similar to `WCS_Action_Scheduler` that relies entirely on Subscription status and which date is being updated (combining the logic of `get_scheduled_action_hook`, `update_status` and `update_date` methods).
- Which conditions we want to check at schedule time vs at the time of sending the notification?
- Do we want to unschedule notifications when store manager disables them? 
  - ~Most likely not, as it's a lot of work to schedule/unschedule all of them. If they're disabled, this will create clutter in the db, though~. Yes, we decided we will do this.

### Side Notes For Testing
- I had to downgrade to node 16 to build it
- I had problems committing as phpcs was giving me errors. I updated `ControlStructureSpacingSniff.php` (converted last argument to array for `ExtraSpaceBeforeCloseParenthesis` and `ExtraSpaceAfterCloseParenthesis` sniffs), otherwise `phpcs` was failing. Perhaps there's a better way to solve this, but I didn't want to spend too much time investigating what's going on and what array should be passed where it gets a string. Updating `squizlabs/php_codesniffer` to 3.10.2 didn't help either. 
  - it seems like [this has been fixed in the latest WP sniffs](https://github.com/WordPress/WordPress-Coding-Standards/blob/develop/WordPress/Sniffs/WhiteSpace/ControlStructureSpacingSniff.php#L169), so maybe updating some composer packages would resolve this (like the old WC sniffs?)


## TODO





## How to test this PR

1. Create a couple of different types of subscriptions: simple, variable, with free trial, with defined end-date/length, with free trial and length
2. Buy these subscriptions, update them, renew them manually and automatically, cancel them. 
3. See if the notifications are being (un)scheduled as expected.
